```
error[E0412]: cannot find type `ServoComputedValues` in this scope
    --> /home/servo/buildbot/slave/linux-dev/build/components/layout/display_list_builder.rs:1875:41
     |
1875 |         let build_local_clip = |style: &ServoComputedValues| {
     |                                         ^^^^^^^^^^^^^^^^^^^ did you mean `ComputedValues`?

   Compiling constellation v0.0.1 (file:///home/servo/buildbot/slave/linux-dev/build/components/constellation)
error: aborting due to previous error

error: Could not compile `layout`.
```

--- a/components/layout/display_list_builder.rs
+++ b/components/layout/display_list_builder.rs
@@ -1871,6 +1871,20 @@ impl FragmentDisplayListBuilding for Fragment {
         let stacking_relative_content_box =
             self.stacking_relative_content_box(stacking_relative_border_box);
 
+        // Adjust the clipping region as necessary to account for `border-radius`.
+        let build_local_clip = |style: &ComputedValues| {
+            let radii = build_border_radius_for_inner_rect(&stacking_relative_border_box, style);
+            if !radii.is_square() {
+                LocalClip::RoundedRect(
+                    stacking_relative_border_box.to_rectf(),
+                    ComplexClipRegion::new(stacking_relative_content_box.to_rectf(),
+                                           radii.to_border_radius(),
+                 ))
+            } else {
+                LocalClip::Rect(stacking_relative_border_box.to_rectf())
+            }
+        };
+
         match self.specific {
             SpecificFragmentInfo::TruncatedFragment(box TruncatedFragmentInfo {
                 text_info: Some(ref text_fragment),
@@ -1916,7 +1930,7 @@ impl FragmentDisplayListBuilding for Fragment {
                 if !stacking_relative_content_box.is_empty() {
                     let base = state.create_base_display_item(
                         &stacking_relative_content_box,
-                        LocalClip::from(clip.to_rectf()),
+                        build_local_clip(&self.style),
                         self.node,
                         self.style.get_cursor(Cursor::Default),
                         DisplayListSection::Content);
@@ -1927,7 +1941,8 @@ impl FragmentDisplayListBuilding for Fragment {
 
                     let size = Size2D::new(item.bounds().size.width.to_f32_px(),
                                            item.bounds().size.height.to_f32_px());
-                    state.iframe_sizes.push((fragment_info.browsing_context_id, TypedSize2D::from_untyped(&size)));
+                    state.iframe_sizes.push((fragment_info.browsing_context_id,
+                                             TypedSize2D::from_untyped(&size)));
 
                     state.add_display_item(item);
                 }
@@ -1937,7 +1952,7 @@ impl FragmentDisplayListBuilding for Fragment {
                 if let Some(ref image) = image_fragment.image {
                     let base = state.create_base_display_item(
                         &stacking_relative_content_box,
-                        LocalClip::from(clip.to_rectf()),
+                        build_local_clip(&self.style),
                         self.node,
                         self.style.get_cursor(Cursor::Default),
                         DisplayListSection::Content);
@@ -1968,7 +1983,7 @@ impl FragmentDisplayListBuilding for Fragment {
 
                 let base = state.create_base_display_item(
                     &stacking_relative_content_box,
-                    LocalClip::from(clip.to_rectf()),
+                    build_local_clip(&self.style),
                     self.node,
                     self.style.get_cursor(Cursor::Default),
                     DisplayListSection::Content);

--- a/components/layout/display_list_builder.rs
+++ b/components/layout/display_list_builder.rs
@@ -2080,7 +2080,6 @@ impl FragmentDisplayListBuilding for Fragment {
                                             stacking_relative_content_box: &Rect<Au>,
                                             text_shadows: &[SimpleShadow],
                                             clip: &Rect<Au>) {
-
         // NB: The order for painting text components (CSS Text Decoration Module Level 3) is:
         // shadows, underline, overline, text, text-emphasis, and then line-through.
 
@@ -2124,7 +2123,7 @@ impl FragmentDisplayListBuilding for Fragment {
         // to back).
 
         // Shadows
-        for shadow in text_shadows.iter().rev() {
+        for shadow in text_shadows.iter() {
             state.add_display_item(DisplayItem::PushTextShadow(box PushTextShadowDisplayItem {
                 base: base.clone(),
                 blur_radius: shadow.blur,

--- a/components/layout/webrender_helpers.rs
+++ b/components/layout/webrender_helpers.rs
@@ -464,7 +464,7 @@ impl WebRenderDisplayItemConverter for DisplayItem {
             DisplayItem::Iframe(ref item) => {
                 let rect = item.base.bounds.to_rectf();
                 let pipeline_id = item.iframe.to_webrender();
-                builder.push_iframe(rect, pipeline_id);
+                builder.push_iframe(rect, Some(item.base.local_clip), pipeline_id);
             }
             DisplayItem::PushStackingContext(ref item) => {
                 let stacking_context = &item.stacking_context;

